In [1]:
import csv
import xlrd
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

# CPS Historical Migration/Geographic Mobility Tables
https://www.census.gov/data/tables/time-series/demo/geographic-mobility/historic.html
- Note: 
    - As of November 2018, these historical tables include corrections to errors that were discovered. 
    - Please replace any previously downloaded historical tables.
- Action:
    - load in tables
    - brief pre-exploration overview analysis (6 total)

In [2]:
# Table A-1. Annual Geographical Mobility Rates, By Type of Movement: 2018
annual_geographical_mobility_rates_type_movement = pd.read_excel( '../data/cps-historical-migration-geographic-mobility/tab-a-1.xls' )

# Table A-2. Annual Inmigration, Outmigration, Net Migration, and Movers from Abroad for Regions: 2018
annual_inmigration_outmigration_net_migration__movers_from_abroad_for_regions = pd.read_excel( '../data/cps-historical-migration-geographic-mobility/tab-a-2.xls' )

# Table A-3. Inmigration, Outmigration, and Net Migration by Metropolitan Status: 2018
inmigration_outmigration__net_migration_metropolitan_status = pd.read_excel( '../data/cps-historical-migration-geographic-mobility/tab-a-3.xls' )

# Table A-4. Geographical Mobility by Tenure: 2018
geographical_mobility_by_tenure = pd.read_excel( '../data/cps-historical-migration-geographic-mobility/tab-a-4.xls' )

# Table A-5. Reason for Move (Specific Categories): 2018
reason_for_move_specific_categories = pd.read_excel( '../data/cps-historical-migration-geographic-mobility/tab-a-5.xls' )

# Table A-6. Distance of Intercounty Move: 2018
distance_of_intercounty_move = pd.read_excel( '../data/cps-historical-migration-geographic-mobility/tab-a-6.xls' )

In [3]:
'''annual_geographical_mobility_rates_type_movement'''

print( annual_geographical_mobility_rates_type_movement.info() )
# annual_geographical_mobility_rates_type_movement.sample( 3 ) 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 186 entries, 0 to 185
Data columns (total 10 columns):
Table with row headers in column A and column headers in rows 5 through 7.    154 non-null object
Unnamed: 1                                                                    137 non-null object
Unnamed: 2                                                                    137 non-null object
Unnamed: 3                                                                    137 non-null object
Unnamed: 4                                                                    138 non-null object
Unnamed: 5                                                                    137 non-null object
Unnamed: 6                                                                    138 non-null object
Unnamed: 7                                                                    137 non-null object
Unnamed: 8                                                                    137 non-null object
Unnamed: 9   

***annual_geographical_mobility_rates_type_movement***
- notes:
    - seems to cover back to 1947
    - needs to be cleaned , pd does not naturally detect formatting 
        - top and bottom
    - varying length of information
        - count (column(s))
            - 154 (0) , 138 (4, 6) , 137 (1, 2, 3, 5, 7, 8, 9)
        - information length is most likely <= 137

In [4]:
'''annual_inmigration_outmigration_net_migration__movers_from_abroad_for_regions'''

# print( annual_inmigration_outmigration_net_migration__movers_from_abroad_for_regions.info() )
annual_inmigration_outmigration_net_migration__movers_from_abroad_for_regions.sample( 3 )

Table with row headers in column A and column headers in row 5.  \
200                                .Movers from abroad                
129                  .Net migration (including abroad)                
11                                                 NaN                

    Unnamed: 1 Unnamed: 2 Unnamed: 3 Unnamed: 4  
200        267        132        451        396  
129         62         79       674*       455*  
11         NaN        NaN        NaN        NaN

***annual_inmigration_outmigration_net_migration__movers_from_abroad_for_regions***
- notes:
    - seems to cover back to 1980
    - needs to be cleaned , pd does not naturally detect formatting 
        - top and bottom

In [5]:
'''inmigration_outmigration__net_migration_metropolitan_status'''

# print( inmigration_outmigration__net_migration_metropolitan_status.info() )
inmigration_outmigration__net_migration_metropolitan_status.sample( 3 )

Table with row headers in column A and column headers in row 5. Unnamed: 1  \
39                                 ..Principal cities                    3206   
42                                                NaN                     NaN   
70                                          ..Suburbs                    5497   

   Unnamed: 2 Unnamed: 3 Unnamed: 4 Unnamed: 5  
39       5420    -2,213*        580    -1,633*  
42        NaN        NaN        NaN        NaN  
70       3081     2,416*        447     2,863*

***inmigration_outmigration__net_migration_metropolitan_status***
- notes:
    - seems to cover back to 1985
    - needs to be cleaned , pd does not naturally detect formatting 
        - top and bottom

In [6]:
'''geographical_mobility_by_tenure'''

# print( geographical_mobility_by_tenure.info() )
geographical_mobility_by_tenure.sample( 3 )

Table with row headers in column A and column headers in rows 5 through 7.  Numbers start in row 10, percentages start in row 184, and frequency distributions start in row 358.  \
273                            .Total, 1 year and over                                                                                                                                 
519                            .Total, 1 year and over                                                                                                                                 
55                           ..In owner-occupied units                                                                                                                                 

    Unnamed: 1 Unnamed: 2 Unnamed: 3 Unnamed: 4 Unnamed: 5 Unnamed: 6  \
273        100    85.2204    14.7796    14.2177    8.52459    5.69313   
519        100        100        100        100        100        100   
55      206520     195798      10722      10419       6959       3460   

    Unnamed: 7 Unnamed: 8 Unnamed: 9  
273    2.89977    2.79336   0.561907  
519        100        100        100  
55        2044       1416        303

***geographical_mobility_by_tenure***
- notes:
    - seems to cover back to 1987
    - needs to be cleaned , pd does not naturally detect formatting 
        - top and bottom

In [7]:
'''reason_for_move_specific_categories'''

# print( reason_for_move_specific_categories.info() )
reason_for_move_specific_categories.sample( 3 )

Table with row headers in columns A and K, and column headers in rows 5 and 6. Numbers start in row 9, percentages start in row 36.  \
26                                          2002-2003                                                                                    
30                                          1998-1999                                                                                    
44                        2009-2010 (2010 controls)/3                                                                                    

   Unnamed: 1 Unnamed: 2 Unnamed: 3 Unnamed: 4 Unnamed: 5 Unnamed: 6  \
26      40093       2679       2814       5055       3546        749   
30      42636       2761       3273       4947       4042        676   
44        100        7.3       11.2       11.8        7.8        2.5   

   Unnamed: 7 Unnamed: 8 Unnamed: 9  ... Unnamed: 12 Unnamed: 13 Unnamed: 14  \
26       1275        101        576  ...        7942        1530        2622   
30       1324        245        842  ...        8861        1657        2539   
44        4.2        0.5        1.3  ...        15.4         4.1        10.8   

   Unnamed: 15 Unnamed: 16 Unnamed: 17 Unnamed: 18 Unnamed: 19 Unnamed: 20  \
26           -        4406        1010         160         565           -   
30           -        4738         841         334         469           -   
44           -         8.7         2.7         0.6         1.5         0.3   

   Unnamed: 21  
26         987  
30        1760  
44         4.4  

[3 rows x 22 columns]

***reason_for_move_specific_categories***
- notes:
    - seems to cover back to 1998
    - needs to be cleaned , pd does not naturally detect formatting 
        - top and bottom

In [8]:
# distance_of_intercounty_move
# print( distance_of_intercounty_move.info() )
distance_of_intercounty_move.sample( 3 )

Table with row headers in column A and column headers in row 5. Numbers start in row 8, percentages start in row 32.  \
20                                          2007-2008                                                                     
9                                                 NaN                                                                     
23                                                NaN                                                                     

   Unnamed: 1 Unnamed: 2 Unnamed: 3 Unnamed: 4 Unnamed: 5  
20      11009       4449       2112       1610       2838  
9         NaN        NaN        NaN        NaN        NaN  
23        NaN        NaN        NaN        NaN        NaN

***distance_of_intercounty_move***
- notes:
    - seems to cover back to 2002
    - needs to be cleaned , pd does not naturally detect formatting 
        - top and bottom

## Aggregate Initial Notes
- ***Dates***
    - first recording year varies by up to 55 years 

| first recording year | dataset |
|- - - -|- - - -|
| 1947 | annual_geographical_mobility_rates_type_movement |
|- - - -|- - - -|
| 1980 | annual_inmigration_outmigration_net_migration__movers_from_abroad_for_regions |
|- - - -|- - - -|
| 1985 | inmigration_outmigration__net_migration_metropolitan_status |
|- - - -|- - - -|
| 1987 | inmigration_outmigration__net_migration_metropolitan_status |
|- - - -|- - - -|
| 1998 | reason_for_move_specific_categories |
|- - - -|- - - -|
| 2002 | distance_of_intercounty_move |

- ***format***
    - first rows of each file seem to serve as information , actual tables start near end of .head()
    - ending rows of each file seem to serve as annotation and disclaimer
- ***actions***:
    - walk though one table at a time
    - let's start with dropping NaN clustered rows

In [9]:
# copy dataframe , generalize name
df = annual_geographical_mobility_rates_type_movement.copy()
# drop columns without any non-NaN values
df = df.dropna(thresh= ( 1 ) )
# return the dataframe to see results
# df
df.sample(7)

Table with row headers in column A and column headers in rows 5 through 7.  \
112                                          2002-2003                           
93                                           2016-2017                           
119                                          1998-1999                           
133                                          1987-1988                           
76                                           1957-1958                           
10                                           2015-2016                           
34                                           1999-2000                           

    Unnamed: 1 Unnamed: 2 Unnamed: 3 Unnamed: 4 Unnamed: 5 Unnamed: 6  \
112        100    85.8106    14.1894    13.7403    8.30561    5.43467   
93         100         89         11       10.6        6.8        3.8   
119        100       84.1       15.9       15.4        9.4        5.9   
133        100       82.2       17.8       17.3         11        6.2   
76      167604     133501      34103      33263      22023      11240   
10      314992     279854      35138      33857      21588      12269   
34      270219     226831      43388      41642      24399      17242   

    Unnamed: 7 Unnamed: 8 Unnamed: 9  
112    2.73503    2.69964   0.449115  
93         2.1        1.7        0.4  
119        3.1        2.8        0.5  
133        3.3          3        0.5  
76        5656       5584        840  
10        7501       4768       1281  
34        8814       8428       1746

- ***notes:***
    - 156 rows remain
    - some column names in row 3 , some column names in row 4 , some column names in row 5
        row 3 = 6 column names
            columns = 0 , 1 , 2 , 3 , 4 , 9
        row 4 = 2 column names
            columns = 5 , 6
        row 5 = 2 column names
            columns = 7 , 8
    - row 4 column 4 and row 5 column 6 contain 'Total'
        not obvious what this is
        but it is not the column name in either case
        in both cases , the column name is one row higher in the same column
    - rows 173 - 185 contain only 1 NaN value
- ***actions:***
    - drop all rows without 2 or more non-NaN values
    - merge rows 3 , 4 , and 5 
        taking out 'Total'

In [10]:
# rename dataframe
df = annual_geographical_mobility_rates_type_movement.copy()

# drop columns with less than 2 non-NaN values
df = df.dropna( thresh= ( 2 ) )

# collect column names as pandas series
col_names_0to4_9 = df.iloc[ 0 ]
col_names_5to6 = df.iloc[ 1 ]
col_names7to8 = df.iloc[ 2 ]

# initialize list for column names
column_names = []

# fill list with first row of column names , no exceptions
for column_name in col_names_0to4_9:    
    column_names.append( column_name )

# set index count to -1 so column 0 == column 0
_ = -1

# replace items 5 and 6 in column_names with column names 5 and 6
for column_name in col_names_5to6:
    _ += 1
    # column_name is string (nan instances = float) and name is not 'Total'   
    if isinstance( column_name , str ) and column_name != 'Total':
        # column_name is aligned to replace a nan         
        if str( column_names[ _ ] ) == 'nan':
            # replace item in column_names with column_name             
            column_names[ _ ] = column_name

# set index count to -1 so column 0 == column 0
_ = -1

# replace items 7 and 8 in column_names with column names 7 and 8
for column_name in col_names7to8:
    _ += 1
    # column_name is string (nan instances = float) and name is not 'Total'   
    if isinstance( column_name , str ) and column_name != 'Total':
        # column_name is aligned to replace a nan         
        if str( column_names[ _ ] ) == 'nan':
            # replace item in column_names with column_name             
            column_names[ _ ] = column_name

# (re)set column names
df.columns = column_names

# drop rows used to extract column names
df = df[3:]

# reset dataframe index
df = df.reset_index( drop=True ) 

# let's see
df.head()

Mobility period Total, 1 year old and over  \
0                          2017-2018                     319319   
1                          2016-2017                     316550   
2                          2015-2016                     314992   
3                          2014-2015                     312295   
4  2013-2014 (98,000 address file)/1                     309601   

  Same residence (non-movers) Total movers  \
0                      286967        32352   
1                      281647        34902   
2                      279854        35138   
3                      275971        36324   
4                      273920        35681   

  Different residence in the United States Same county Different county  \
0                                    31186       19852            11334   
1                                    33647       21614            12033   
2                                    33857       21588            12269   
3                                    34651       22905            11746   
4                                    34548       23436            11112   

  Same state Different state Movers from abroad  
0       6496            4838               1165  
1       6668            5366               1255  
2       7501            4768               1281  
3       6653            5093               1673  
4       6446            4666               1133

- ***as of now***:
    - removed rows with NaN values
    - extracted, sorted, and set column names
- ***next (action)***:
    - we need to clean up the 'Mobility period' column
        - samples from current:
            2017-2018
            2013-2014 (98,000 address file)/1
            2011-2012
            2010-2011 (2010 controls)/2
            2000-2001 (SCHIP, 2000 controls)/5
            1970-1971/8
        - what we want:
            - 2018 << 2017-2018
            - 2014 << 2013-2014 (98,000 address file)/1
                requires further investigation
            - 2012 << 2011-2012
            - 2011 << 2010-2011 (2010 controls)/2
                requires further investigation
            - 2001 << 2000-2001 (SCHIP, 2000 controls)/5
                requires further investigation
            - 1971 << 1970-1971/8

In [24]:
# examine data in 'Mobility period'
# _ = 0
# for instance in df[ 'Mobility period' ]:
#     # print( _ , instance )
#     _ += 1
    
# selected outputs (to avoid long out)
'''
# non-uniform , non-repeated
    3 2014-2015
    4 2013-2014 (98,000 address file)/1
    5 2012-2013
    
# repeats w/ controls , no uniform representation
    6 2011-2012
    7 2010-2011 (2010 controls)/2
    8 2010-2011 (2000 controls)/3
    9 2009-2010 (2010 controls)/2
    10 2009-2010 (2000 controls)/3
    11 2008-2009

    18 2001-2002
    19 2000-2001 (2000 controls)/3
    20 2000-2001 (SCHIP, 2000 controls)/5
    21 2000-2001 (1990 controls)/6
    22 1999-2000
    
    28 1993-1994
    29 1992-1993 (1990 controls)/6
    30 1992-1993 (1980 controls)/7
    31 1991-1992
    
# lack of info from 70s 
    41 1981-1982
    42 1980-1981
    43 1975-1976/8
    44 1970-1971/8
    45 1969-1970
    46 1968-1969
    47 1967-1968
    
# in chronological order until row 68 (then repeats)
    66 1948-1949
    67 1947-1948
    68 2017-2018
    69 2016-2017
'''
_ = 68
# make sure rows 0-67 are the same as 68-135
# for row in df[ 'Mobility period' ][ 0 : 68 ]:
#     _ += 1
#     if row != df[ 'Mobility period' ][ _ ]:
#         print(row , df[ 'Mobility period' ][ _ ])
_ = 0
for row in df[ 'Mobility period' ]:
    print(_,row)
    _+=1

0 2017-2018
1 2016-2017
2 2015-2016
3 2014-2015
4 2013-2014 (98,000 address file)/1
5 2012-2013
6 2011-2012
7 2010-2011 (2010 controls)/2
8 2010-2011 (2000 controls)/3
9 2009-2010 (2010 controls)/2
10 2009-2010 (2000 controls)/3
11 2008-2009
12 2007-2008
13 2006-2007
14 2005-2006
15 2004-2005/4
16 2003-2004
17 2002-2003
18 2001-2002
19 2000-2001 (2000 controls)/3
20 2000-2001 (SCHIP, 2000 controls)/5
21 2000-2001 (1990 controls)/6
22 1999-2000
23 1998-1999
24 1997-1998
25 1996-1997
26 1995-1996
27 1994-1995
28 1993-1994
29 1992-1993 (1990 controls)/6
30 1992-1993 (1980 controls)/7
31 1991-1992
32 1990-1991
33 1989-1990
34 1988-1989
35 1987-1988
36 1986-1987
37 1985-1986
38 1984-1985
39 1983-1984
40 1982-1983
41 1981-1982
42 1980-1981
43 1975-1976/8
44 1970-1971/8
45 1969-1970
46 1968-1969
47 1967-1968
48 1966-1967
49 1965-1966
50 1964-1965
51 1963-1964
52 1962-1963
53 1961-1962
54 1960-1961
55 1959-1960
56 1958-1959
57 1957-1958
58 1956-1957
59 1955-1956
60 1954-1955
61 1953-1954
62 19

In [ ]:
# determine type of data in 'Mobility period' 
def test_type( dataframe , test_type ):
    '''
    input) dataframe w/ specified column 
        - e.g. df['specified column']
    input) type wanting to test
        - e.g. str 
    
    output) count of instances which are not type == test_type
    '''
    non_test_type = 0
    for _ in dataframe:
        if isinstance( _ , test_type ) == False:
            non_test_type += 1
    return non_test_type


test_000 = test_type( dataframe=df[ 'Mobility period' ] , test_type=str )
test_000

- ***as test_000 returns 0***:
    - the type of all datapoints in 'Mobility period' is str (string)
    - we will not need to change any points to have uniformity   
- ***next up (action)***:
    - clear low-hanging fruit
    - change the simple ranges to int for set year
        - e.g. change "2017-2018" to "2018"

In [ ]:
# identify changes which can be implimented without further investigation
mobility_period = df[ 'Mobility period' ]
simple_year_ranges_count_ = 0
confirmed_simple_year_ranges_count_ = 0

# count how many simple-fix years we have
for datapoint in mobility_period:
    # len('2012-2013') == 9 , check datapoint is formatted as expected     
    if len( datapoint ) == 9 and len( datapoint.split( '-' ) ) == 2:
        # pretty sure we know what's going on but will investigate further         
        simple_year_ranges_count_ += 1
        # first 4 characters of string , can be converted to int
        first_year = int( datapoint[ :4 ] )
        # last 4 characters of string , can be converted to int
        second_year = int( datapoint[ -4: ] )
        # first_year and last_year are post-1940 (because expected), and year range is 1 year 
        if first_year and second_year > 1940 and second_year - first_year == 1:
            # the datapoint is within our risk threshold
            confirmed_simple_year_ranges_count_ += 1

# if everything we thought was what we think it is 
if simple_year_ranges_count_ == confirmed_simple_year_ranges_count_:
    print(f'total: {confirmed_simple_year_ranges_count_}\nratio: {confirmed_simple_year_ranges_count_/len(mobility_period)}')

- ***notes***:
    - know count of lowest fruit to be 110 ( > 80.88% , not bad)
    - just confirmed status of 110 points to be (at least) similar enough to extract years
    - knowing simple_year_ranges_count_ allows the conformation of outcome via count compairson 
- ***action***:
    - convert those 110 year ranges to years represented by the 2nd year in each range

In [ ]:
less_low_mobility_periods = []
# so row 0 == row 0
_ = -1
# little bit of ctrl + c , little bit of ctrl + v
for datapoint in mobility_period:
    _ += 1
    # len('2012-2013') == 9 , check datapoint is formatted as expected     
    if len( datapoint ) == 9 and len( datapoint.split( '-' ) ) == 2:
        # replace datapoint w/ int of last 4 characters of string 
        mobility_period[ _ ] = int( datapoint[ -4: ] )
    else:
        # keep what's already there
        less_low_mobility_periods.append( datapoint )

mobility_period.sample(7)

- ***notes***:
    - we have converted the low hanging fruit
- ***actions***:
    - now on to exploring the other 26 rows

In [ ]:
# print them out , let's see what we can see
_ = 0
for row in df['Mobility period']:
    _ += 1    
    if not isinstance( row , int ):
        if _ < 10:
            print( str(_) + '  |  ' + row )
        elif _ < 100:
            print( str(_) + ' |  ' + row )
        else:
            print( str(_) + '|  ' + row )

- ***notes***
    - 6 rows formatted almost like majority (new lowest)
        - all have either /4 or /8 at the end of otherwise standard string
            - e.g. standard 
                - 2004-2005
            - observation rows 
                - 16 |  2004-2005/4
                - 44 |  1975-1976/8
                - 45 |  1970-1971/8
                - 84 |  2004-2005/4
                - 112|  1975-1976/8
                - 113|  1970-1971/8
        - remarks before reading documentation
            - identical instances
                - rows 16 and 84
                - rows 44 and 112
                - rows 45 and 113
            - rows are not ordered by year
                - unless years reoccour/loop through table
                - or otherwise
    - 16 rows are in the next new lowest format
        - all contain ({year} controls)/{num} with year and num in consistent pairs
            - (year / num) pairs = # occurrences
                - (2010 / 2) = 4
                - (2000 / 3) = 6
                - (1990 / 6) = 4
                - (1980 / 7) = 2
        - they occour in pattern twice:
            - 4 uninterrupted
            - 1
            - 3 uninterrupted
        - sample rows
            - 8  |  2010-2011 (2010 controls)/2
            - 9  |  2010-2011 (2000 controls)/3
            - 10 |  2009-2010 (2010 controls)/2
            - 11 |  2009-2010 (2000 controls)/3
        - unknown analysis
            - seems to be game settings for different censuses 
            - difficult to connect the /2,/3 and /6,/7 relationship
                - if was dependent on decade order , possibles would be straight ordered 
                    - e.g. /2, /3, /4, /5 or /1, /3, /5, /7
                    
    - final 4 rows are longer and are 2 duplicate pairs
        - default format + ending either (98,000 address file)/1 or (SCHIP, 2000 controls)/5
            - longest rows
                - 5  |  2013-2014 (98,000 address file)/1
                - 21 |  2000-2001 (SCHIP, 2000 controls)/5
                - 73 |  2013-2014 (98,000 address file)/1
                - 89 |  2000-2001 (SCHIP, 2000 controls)/5
        - pre document remarks
            - first notice of /1 
            - return use of /5 
                - do other /n have additional possibile (paramaters)
    - blind analysis
        - year range
                - min: 1970-1971 (1971 in complete df)
                - max: 2013-2014 (2014 in complete df)
            - less range than full representation of column
                - min: 1948 (1947-1948 in origional df)
                - max: 2018 (2017-2018 in origional df)
            - why?
                - recent changes to census
                    - collection
                    - questions
                    - coverage
                - recent increase in census related surveys
                    - housing surveys
                    - business surveys
                        - we will be exploring these after fitting gen_000 model
                        - these are also how this data is presented for non-decade years
        - lot of possible meanings for non-conforming objects
                - politics..
                - changing standards
                - technological evolution
                - etc...
                    - storage , collection , methods , etc... 
            - full-court lobs
                - something to do with immigrants
                - evolution of process 
                - expansion of non-physical opportunities 
        - there's got to be a better way to extract dataframes from excel files
            - likely convertable, but annotations likely direct either universe to same result 
- ***actions***
    - go find out what these non-standard occurrences mean

In [ ]:
# make another copy of origional dataframe (this is why we copy)
info_df = annual_geographical_mobility_rates_type_movement.copy()

# function to make NaN dataframe
row_nans = lambda df: df[ df.isnull().any( axis=1 ) ]

# identify the rows with information (basically, keep the NaNs this time)
nan_df = row_nans( info_df )

# peak some rows with NaN values 
nan_df.head()

In [ ]:
# noted column 0 name 
column_names = []

# make it slightly educational
for _ in nan_df[ :0 ]:
    # each _ is a string
    if isinstance( _ , str ):
        # add the string to column_names
        column_names.append( _ )
        
# this is the education
print( f'len(nan_df[:0])=={ len( nan_df[ :0 ] ) }\ntype(nan_df[:0])=={ type( nan_df[ :0 ] ) }\n'
      # The economy, stupid. 
       f'len(column_names)=={ len( column_names ) }\ntype(column_names={ type( column_names ) }' )

In [ ]:
# we can find row headers in column A
if column_names[0] != 'Table with row headers in column A and column headers in rows 5 through 7.':
    raise AssertionError('')

# function to return non-NaN rows 
non_nans = lambda df: df.dropna()

# identify non-NaN rows from column A (the first column)
column_a = non_nans( nan_df[ 'Table with row headers in column A and column headers in rows 5 through 7.' ] )

# let's see
column_a

In [ ]:
# identify information of interest
indicators_of_interest = [ '1/' , '2/' , '3/' , '4/' , '5/' , '6/' , '7/' , '8/' ]

# make copy
copy_column_a = column_a.copy()

# return rows from copy_column_a which contain an indicator from indicators_of_interest
rows_of_interest = [ row for indicator in indicators_of_interest for row in copy_column_a if indicator in row ]

# examine the lowest (remember: /4 and /8)
for row in rows_of_interest:
    if '4/' in row or '8/' in row:
        print( row + '\n' )

https://www.census.gov/population/www/socdemo/CPSnote.pdf 

notes:
    - f

In [ ]:
plt.matshow(annual_geographical_mobility_rates_type_movement.corr())
plt.show()